### DISCUSSION TITLE
Cannot authenticate using AWS IAM RDS Auth

### DISCUSSION THREAD

**Discussion by AcerLaze** [theory/sqitch#819] [Q&A]
Hello, I want to set up a jenkins pipeline to automate DB deployment.

I have set up all the required settings on the aws & managed to retrieved the token using this [guide from aws](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.Connecting.AWSCLI.html).
If I understand it correctly I just need to put the token in the env $SQITCH_PASSWORD and then run the command deploy
`sqitch deploy db:mysql://{$dbuser}@{$dbHost}:3306/${dbName}`

But when I tried to do that I get access denied.

Am I missing something or are there something wrong with how I tried to authenticate to the database ?

For more information :
- I'm using sqitch from docker image
- DB Version : Aurora MySQL 3.04.1 (compatible with MySQL 8.0.28)

**Comment by theory** [comment]
I have no idea, but can [ask around](https://xoxo.zone/@theory/112129328817694069). Meanwhile could you post your pipeline file so others can review how things are set up?

**Comment by AcerLaze** [comment]
Sure, here's my pipeline file content.

```@Library('pipeline-library')
import com.sec.devops.AwsUtil

def gitUrl = env.GIT_URL

pipeline {
  agent {
    node {
      label params.AGENT_LABEL
    } 
  }
  stages{
    stage('Cleanup') {
      steps {
        cleanWs()
      }
    }
    stage('Checkout') {
      steps{
        script {
          def repoUrl = gitUrl + params.ORG_NAME +'/'+ params.REPO_NAME
          git url: repoUrl, branch: params.BRANCH, credentialsId: env.GIT_CREDENTIALS_ID
        }
      }
    }
    stage('Setup Sqitch') {
      steps{
        script {
          sh "docker pull sqitch/sqitch"
          sh "curl -L https://git.io/JJKCn -o sqitch && chmod +x sqitch"
        }
      }
    }
    stage('Assume Role') {
      steps {
        script {
          def roleArn = "arn:aws:iam::${params.AWS_ACCOUNT_ID}:role/${params.IAM_ROLE}"
          def service = env.SERVICE_NAME
          def sessionName = env.AWS_SESSION_NAME
          def stageEnv = params.STAGE_ENV
          def awsUtil = new AwsUtil(this)
          awsUtil.assume_role(stageEnv, service, roleArn, sessionName)
        }
      }
    }
    stage('Deploy Changes') {
      steps {
        script {
          def dbHostname = params.DB_HOSTNAME
          def dbPort = params.DB_PORT
          def dbName = params.DB_NAME
          def region = params.REGION
          def dbUser = params.DB_USER
          env.SQITCH_PASSWORD = sh(
            script: "aws rds generate-db-auth-token --hostname ${dbHostname} --port ${dbPort} --region ${region} --username ${dbUser}",
            returnStdout: true
          )
          sh "./sqitch deploy db:mysql://${params.DB_HOSTNAME}:${params.DB_PORT}/${dbName}"
        }
      }
    }
  }
}
```

I don't have the access to view the AwsUtil import so I can't share it, but basically it only used to simplify assume role process on aws.
The assume role part works perfectly, because i can generate the token & set it on `SQITCH_PASSWORD`

**Reply by theory** [reply]
Try adding the username to the URI:

```
sh "./sqitch deploy db:mysql://${params.DB_USER}@${params.DB_HOSTNAME}:${params.DB_PORT}/${dbName}"
```

**Reply by AcerLaze** [reply]
Whoops seems I forgot to include the username when updating the script before paste it here.
In the original script it already included the username and it still doesn't work.

**Reply by AcerLaze** [reply]
Yesterday I've tried it a bit more and found out using normal username & password I can connect to rds just fine.
So i think the main issue is with the token generated by aws

**Reply by theory** [reply]
Yeah, not at all sure one can use the token for the password, but I am not familiar with it, so maybe? Are there any special characters that might need escaping, or white space, requiring the environment variable to be quoted?

**Reply by AcerLaze** [reply]
What kind of special character that usually needed to be escaped ?

The token generated by aws looks like an url with some query parameter.
Aws doc have the example token 
`rdsmysql.123456789012.us-west-2.rds.amazonaws.com:3306/?Action=connect&DBUser=jane_doe&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Expires=900...`

**Reply by theory** [reply]
In a shell you should wrap values in quotation marks. I don't recall if that's necessary in Jenkins pipelines, though, but the question mark and ampersand would definitely cause issues, because they have explicit meanings as operators in a shell context.

**Reply by AcerLaze** [reply]
I've checked the `SQITCH_PASSWORD` by printing the env before running the sqitch deploy and the value is correct.
Also already tried setting the value manually wrapped in quotation marks and still get access denied.

**Reply by AcerLaze** [reply]
So far from my testing, access denied only occurred when using token generated via IAM, but I can connect using db user & password.
I think I'm gonna give up trying to create sqitch jenkins pipeline with iam authentication.
Instead create an user on the db and use that to deploy changes.

**Reply by theory** [reply]
It looks like there's [documentation](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.Connecting.AWSCLI.html) for using a token with the `mysql` CLI, but not with SDKs like DBD::mysql, which Sqitch uses to connect to the registry. It may be that there's an incantation that would work by adding query parameters for the SSL CA (`mysql_ssl_ca_file`) and enabling cleartext auth, but I don't know what they are, and they might require modifications to [App::Sqitch::Engine::mysql](https://github.com/sqitchers/sqitch/blob/5522821/lib/App/Sqitch/Engine/mysql.pm#L189-L201) to work. And I don't see a likely candidate for the cleartext auth plugin in the [DBD::mysql docs](https://metacpan.org/pod/DBD::mysql#connect).

**Reply by girzel** [reply]
I believe this is an escaping issue. I've had to run the IAM tokens through `sed 's/:/\\:/g'` before they were usable as values for PGPASSWORD (and I assume SQITCH_PASSWORD as well).
